File: studio/AuraStudio/Tests/AuraStudioTests/USBDeviceIdentityTests.swift
import XCTest
@testable import AuraStudio

/// ST-016: la unica lectura real de "que firmware corre" es lo que el
/// propio firmware anuncia por USB. Estos tests fijan la clasificacion
/// contra los descriptores reales (iPod del dueño en modo disco de Apple,
/// leidos con `ioreg`; y los de Rockbox, `usb_core.c` del firmware).
final class USBDeviceIdentityTests: XCTestCase {
    func testAppleDiskModeIsClassifiedAsApple() {
        XCTAssertEqual(RunningFirmware.classify(vendorName: "Apple Inc.", productName: "iPod"), .apple)
    }

    func testRockboxDescriptorsAreClassifiedAsRockboxFamily() {
        XCTAssertEqual(RunningFirmware.classify(vendorName: "Rockbox.org", productName: "Rockbox media player"),
                       .rockboxFamily)
        // Basta con que una de las dos cadenas lo diga.
        XCTAssertEqual(RunningFirmware.classify(vendorName: "", productName: "Rockbox media player"),
                       .rockboxFamily)
    }

    func testAnythingElseIsUnknownNeverGuessed() {
        XCTAssertEqual(RunningFirmware.classify(vendorName: "", productName: ""), .unknown)
        XCTAssertEqual(RunningFirmware.classify(vendorName: "Apple Inc.", productName: "iPad"), .unknown)
        XCTAssertEqual(RunningFirmware.classify(vendorName: "Ugreen", productName: "USB3 Hub"), .unknown)
    }

    func testIPodClassicVIDPIDIsRecognised() {
        let ipod = USBDeviceIdentity(vendorName: "Apple Inc.", productName: "iPod", serialNumber: "000A270013923F13",
                                     vendorID: 0x05AC, productID: 0x1261)
        XCTAssertTrue(ipod.isIPodClassicUSB)
        XCTAssertEqual(ipod.runningFirmware, .apple)
    }

    /// Un iPad tambien es 0x05AC -- el PID es lo que decide.
    func testOtherAppleDevicesAreNotIPodClassic() {
        let ipad = USBDeviceIdentity(vendorName: "Apple Inc.", productName: "iPad", serialNumber: nil,
                                     vendorID: 0x05AC, productID: 0x12AB)
        XCTAssertFalse(ipad.isIPodClassicUSB)
    }

    func testRockboxKeepsAppleVIDPIDSoIdentityStillMatches() {
        let running = USBDeviceIdentity(vendorName: "Rockbox.org", productName: "Rockbox media player",
                                        serialNumber: nil, vendorID: 0x05AC, productID: 0x1261)
        XCTAssertTrue(running.isIPodClassicUSB)
        XCTAssertEqual(running.runningFirmware, .rockboxFamily)
    }
}

File: studio/windows/tests/AuraStudio.Core.Tests/USBDeviceIdentityTests.cs
using AuraStudio.Core;
using Xunit;

namespace AuraStudio.Core.Tests;

/// <summary>
/// ST-016: la única lectura real de "qué firmware corre" es lo que el propio
/// firmware anuncia por USB. Estos tests fijan la clasificación contra los
/// descriptores reales (iPod del dueño en modo disco de Apple y los de
/// Rockbox).
/// </summary>
public class USBDeviceIdentityTests
{
    [Fact]
    public void AppleDiskMode_IsClassifiedAsApple()
    {
        Assert.Equal(RunningFirmware.Apple, RunningFirmware.Classify("Apple Inc.", "iPod"));
    }

    [Fact]
    public void RockboxDescriptors_AreClassifiedAsRockboxFamily()
    {
        Assert.Equal(RunningFirmware.RockboxFamily, RunningFirmware.Classify("Rockbox.org", "Rockbox media player"));
        // Basta con que una de las dos cadenas lo diga.
        Assert.Equal(RunningFirmware.RockboxFamily, RunningFirmware.Classify("", "Rockbox media player"));
    }

    [Fact]
    public void AnythingElse_IsUnknownNeverGuessed()
    {
        Assert.Equal(RunningFirmware.Unknown, RunningFirmware.Classify("", ""));
        Assert.Equal(RunningFirmware.Unknown, RunningFirmware.Classify("Apple Inc.", "iPad"));
        Assert.Equal(RunningFirmware.Unknown, RunningFirmware.Classify("Ugreen", "USB3 Hub"));
    }

    [Fact]
    public void IPodClassicVIDPID_IsRecognised()
    {
        var ipod = new USBDeviceIdentity("Apple Inc.", "iPod", "000A270013923F13", 0x05AC, 0x1261);
        Assert.True(ipod.IsIPodClassicUSB);
        Assert.Equal(RunningFirmware.Apple, ipod.RunningFirmware);
    }

    /// Un iPad también es 0x05AC — el PID es lo que decide.
    [Fact]
    public void OtherAppleDevices_AreNotIPodClassic()
    {
        var ipad = new USBDeviceIdentity("Apple Inc.", "iPad", null, 0x05AC, 0x12AB);
        Assert.False(ipad.IsIPodClassicUSB);
    }

    [Fact]
    public void RockboxKeepsAppleVIDPID_SoIdentityStillMatches()
    {
        var running = new USBDeviceIdentity("Rockbox.org", "Rockbox media player", null, 0x05AC, 0x1261);
        Assert.True(running.IsIPodClassicUSB);
        Assert.Equal(RunningFirmware.RockboxFamily, running.RunningFirmware);
    }
}
